accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,375 @@
+# Dart Static Extension Methods Design
+lrn@google.com
+
+This is a design document for *static extension methods* for Dart. This document describes the most basic variant of the feature, then lists a few possible variants or extensions.
+
+The design of this feature is kept deliberately simple.
+
+## What Are Static Extension Methods
+
+Dart classes have virtual methods. An invocation like `thing.doStuff()` will invoke the virtual `doStuff` method on the object denoted by `thing`. The only way to add methods to a class is to modify the class. If you are not the author of the class, you have to use static helper functions instead of methods, so `doMyStuff(thing)` instead of `thing.doMyStuff()`. That's acceptable for a single function, but in a larger chain of operations, it becomes overly cumbersome. Example:
+
+```dart
+doMyOtherStuff(doMyStuff(something.doStuff()).doOtherStuff())
+```
+
+That code is so much less readable than
+
+```dart
+something.doStuff().doMyStuff().doOtherStuff().doMyOtherStuff()
+```
+
+So, primarily for readability reasons, static extension methods will allow you to add "extension methods", which are really just static functions, to existing types, and allow you to call those methods as if they were actual methods, using `.`.
+
+The extension methods are *static*, which means that we use the static type of an expression to figure out which method to call, and that also means that static extension methods are not virtual methods.
+
+The methods we allow you to add this way are normal methods, operators, getter and setters.
+
+## Declaring Static Extension Methods
+
+### Syntax
+
+A static extension of a type is declared using syntax like:
+
+```dart
+extension MyFancyList<T> on List<T> {
+  int get doubleLength => this.length * 2;
+  List<T> operator-() => this.reversed.toList();
+  List<List<T>> split(int at) => 
+      <List<T>>[this.sublist(0, at), this.sublist(at)];
+  List<T> mapToList<R>(R Function(T) convert) => this.map(convert).toList();
+}
+```
+
+More precisely, an extension declaration is a top-level declaration with a grammar similar to:
+
+```ebnf
+<extension> ::= 
+  `extension' <identifier> <typeParameters>? `on' <type> `?'? `{'
+     memberDeclaration*
+  `}'
+```
+
+Such a declaration introduces its *name* (the identifier) into the surrounding scope. The name does not denote a type, but it can be used to denote the extension itself in various places. The name can be hidden or shown in `import` or `export` declarations.
```

Comment:
What do you think about making the name optional?  I really do think there is a good use case for file local extensions, where I want to define a bunch of helper operations on some types, but I don't plan to export them.  Having to pick a meaningless name that doesn't conflict with anything else in the file or in the imported extensions seems like pointless work:  why not just allow me to say:

```
extension on int {
  // My int helper methods here
}
extension on String {
  // My String helper methods here
}
```

Reply:
I'm not sure how often library-private extensions occur. I'd wait and see if there is a market for this feature. All you need to do is write `_1` instead if you really don't care about the name.

Reply:
> All you need to do is write `_1`

I know, and that's horrid.  We can certainly wait on allowing it, but there's an opportunity cost to waiting.  If we don't do this now, odds are very high we'll never do it.  Every change has a fixed cost.

I would expect library private extensions to be approximately as common as library private functions/static methods, since that's what we're doing: giving people a way to write functions that have a nice call syntax.

Reply:
Yes, unnamed extensions would not be exported.  The semantics would be that the compiler invents a unique library private name to fill in for the missing name.